Pearson-Advance/openedx-lti-tool-plugin
# PADV-263: Implement LTI 1.3 authentication

## Ticket

https://agile-jira.pearson.com/browse/PADV-263

## Description

This PR adds authentication to the LTI 1.3 plugin, this includes required dpeendencies, model authentication backend, required LtiProfile custom save method, model fields, LTI 1.3 login endpoint, JWKS endpoint and launch endpoint with only user authentication.

## Type of Change

- [x] Add edx-opaque-keys to base requirements.
- [x] Add LTI 1.3 authentication ModelBackend.
- [x] Add LTI 1.3 profile model fields, profile_url and custom save method.
- [x] Add LTI 1.3 profile model manager methods.
- [x] Add unquote_slashes function from edx-platform.
- [x] Add function to parse course and usage keys using edx-opaque-keys.
- [x] Add model migrations
- [x] Add requires_lti_enabled view method decorator.
- [x] Implement LTI 1.3 login endpoint.
- [x] Implement LTI 1.3 public JWKS endpoint.
- [x] Implement LTI 1.3 launch endpoint with only authentication.

## Testing:

- Run the LMS: `make dev.up.lms`.
- Install `openedx_lti_tool_plugin` on the LMS.
- Run migrations on the LMS: `./manage lms migrate`
- Add required settings to LMS.
	
```
    # Enable LTI Tool Provider Plugin.
    FEATURES['ENABLE_LTI_PROVIDER_PLUGIN'] = True

    if FEATURES.get('ENABLE_LTI_PROVIDER_PLUGIN'):
        AUTHENTICATION_BACKENDS.append('openedx_lti_tool_plugin.auth.LtiAuthenticationBackend')
```

- Run ngrok on the LMS: `ngrok http 18000`.
- Go to https://saltire.lti.app/platform.
- Once logged, on the sidebar go to "Security Model".
- Set "LTI Version" to "1.3.0".
- Set "Message URL" to the launch URL on the LMS, Example: https://2106-190-142-38-224.ngrok.io/openedx_lti_tool_plugin/1.3/launch/?course_id=course-v1%3AedX%2BDemoX%2BDemo_Course&usage_id=block-v1%3AedX%2BDemoX%2BDemo_Course%2Btype%40sequential%2Bblock%40edx_introduction (Note: URL parameters must be URL encoded).
- Leave fields on "Platform Details" as they are, these are the settings for this platform identity (consuming client).
- On "Tool Details" set "Initiate login URL" to our LTI tool URL, Example: https://2106-190-142-38-224.ngrok.io/openedx_lti_tool_plugin/1.3/login/
- Add the URL set on "Message URL" to "Redirection URI(s)".
- Set "Public keyset URL" to our LTI tool JWKS endpoint, Example: https://2106-190-142-38-224.ngrok.io/openedx_lti_tool_plugin/1.3/pub/jwks/
- Save settings using the "Save" button on the top left corner.
- Go to the LMS admin panel.
- Go to PyLTI 1.3 Tool Config › Lti 1.3 tool keys and add a key.
- On the "Private key" field, copy the private key shown on the "Platform Details" section in https://saltire.lti.app/platform.
- On the "Public key" field, copy the public key shown on the "Platform Details" section in https://saltire.lti.app/platform.
- Go to PyLTI 1.3 Tool Config › Lti 1.3 tool and add a tool config.
- Set "Issuer" to the "Platform/Issuer ID" value shown on the "Platform Details" section in https://saltire.lti.app/platform.
- Set "Client id" to the "Client ID" value shown on the "Platform Details" section in https://saltire.lti.app/platform.
- Set "Auth login url" to the "Authentication request URL" value shown on the "Platform Details" section in https://saltire.lti.app/platform.
- Set "Auth token url" to the "Access Token service URL" value shown on the "Platform Details" section in https://saltire.lti.app/platform.
- Set "Key set url" to the https://saltire.lti.app/platform/jwks (Note: Make sure that you don't create a new private key on the platform emulator, there is a bug and the public key set they assign you doesn't work, so you need to use the default JWKS set with the default private and public keys).
- Set "Tool key" to the LTI 1.3 tool key you created previously.
- Set "Deployment ids" to the "Deployment ID" value shown on the "Platform Details" section in https://saltire.lti.app/platform.
- Go back to https://saltire.lti.app/platform.
- Click on the "Connect" button on the top right corner.
- A new tab with the LTI user authenticated information should be displayed.

## Reviewers
